### PR TITLE
Add PrestoSparkShuffleDescriptor for carrying over shuffle info from Spark

### PR DESCRIPTION
**Pull request by tanjialiang**
1. Added PrestoSparkShuffleDescriptor and its children classes. They are used to carry over shuffle information in Spark language and later used for native processes shuffles.
2. Refactoring on PrestoSparkLocalShuffleReadInfo/PrestoSparkLocalShuffleWriteInfo to be consistent to C++ structures in native execution codebase.
```
== NO RELEASE NOTE ==
```

### CODE REVIEW

**Review by amitkdutta — APPROVED** [review]
Looks good.

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/shuffle/PrestoSparkLocalShuffleInfoTranslator.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    @Inject
+    public PrestoSparkLocalShuffleInfoTranslator(
+            JsonCodec<PrestoSparkLocalShuffleReadInfo> readInfoJsonCodec,
+            JsonCodec<PrestoSparkLocalShuffleWriteInfo> writeInfoJsonCodec)
+    {
+        try {
+            this.localShuffleRootPath = Files.createTempDirectory("local_shuffle").toAbsolutePath().toString();
+        }
+        catch (IOException e) {
+            throw new RuntimeException(e);
```
Comment (arunthirupathi) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/execution/shuffle/PrestoSparkLocalShuffleInfoTranslator.java:46]: Wrap this is UncheckedIOException https://docs.oracle.com/javase/8/docs/api/java/io/UncheckedIOException.html

Also add some descriptive message so the failure is more readable.

File: presto-spark-base/src/test/java/com/facebook/presto/spark/execution/TestBatchTaskUpdateRequest.java [hunk 2]
```diff
@@ -68,10 +69,9 @@
     public void testJsonConversion()
             throws Exception
     {
-        PrestoSparkShuffleInfoSerializer shuffleInfoSerializer = new PrestoSparkLocalShuffleInfoSerializer(PRESTO_SPARK_LOCAL_SHUFFLE_READ_INFO_JSON_CODEC, PRESTO_SPARK_LOCAL_SHUFFLE_WRITE_INFO_JSON_CODEC);
-        PrestoSparkShuffleReadInfo readInfo = new PrestoSparkLocalShuffleReadInfo(0, 0, 0, "/dummy/read/path");
-        byte[] serializedReadInfo = shuffleInfoSerializer.serializeReadInfo(readInfo);
-        String stringSerializedReadInfo = PRESTO_SPARK_LOCAL_SHUFFLE_READ_INFO_JSON_CODEC.toJson(PRESTO_SPARK_LOCAL_SHUFFLE_READ_INFO_JSON_CODEC.fromBytes(serializedReadInfo));
+        PrestoSparkLocalShuffleInfoTranslator shuffleInfoTranslator = new PrestoSparkLocalShuffleInfoTranslator(PRESTO_SPARK_LOCAL_SHUFFLE_READ_INFO_JSON_CODEC, PRESTO_SPARK_LOCAL_SHUFFLE_WRITE_INFO_JSON_CODEC);
```
Comment (arunthirupathi) [inline presto-spark-base/src/test/java/com/facebook/presto/spark/execution/TestBatchTaskUpdateRequest.java:72]: parameter formatting.

File: presto-spark-base/src/test/java/com/facebook/presto/spark/execution/TestBatchTaskUpdateRequest.java [hunk 3]
```diff
@@ -115,13 +115,11 @@ public void testJsonConversion()
...
     {
-        PrestoSparkShuffleInfoSerializer shuffleManager = new PrestoSparkLocalShuffleInfoSerializer(PRESTO_SPARK_LOCAL_SHUFFLE_READ_INFO_JSON_CODEC, PRESTO_SPARK_LOCAL_SHUFFLE_WRITE_INFO_JSON_CODEC);
-        PrestoSparkShuffleReadInfo readInfo = new PrestoSparkLocalShuffleReadInfo(0, 0, 0, "/dummy/read/path");
-        PrestoSparkShuffleWriteInfo writeInfo = new PrestoSparkLocalShuffleWriteInfo(1, 1, "/dummy/write/path");
-        byte[] serializedReadInfo = shuffleManager.serializeReadInfo(readInfo);
-        byte[] serializedWriteInfo = shuffleManager.serializeWriteInfo(writeInfo);
-        String stringSerializedReadInfo = PRESTO_SPARK_LOCAL_SHUFFLE_READ_INFO_JSON_CODEC.toJson(PRESTO_SPARK_LOCAL_SHUFFLE_READ_INFO_JSON_CODEC.fromBytes(serializedReadInfo));
-        String stringSerializedWriteInfo = PRESTO_SPARK_LOCAL_SHUFFLE_WRITE_INFO_JSON_CODEC.toJson(PRESTO_SPARK_LOCAL_SHUFFLE_WRITE_INFO_JSON_CODEC.fromBytes(serializedWriteInfo));
+        PrestoSparkLocalShuffleInfoTranslator shuffleTranslator = new PrestoSparkLocalShuffleInfoTranslator(PRESTO_SPARK_LOCAL_SHUFFLE_READ_INFO_JSON_CODEC, PRESTO_SPARK_LOCAL_SHUFFLE_WRITE_INFO_JSON_CODEC);
```
Comment (arunthirupathi) [inline presto-spark-base/src/test/java/com/facebook/presto/spark/execution/TestBatchTaskUpdateRequest.java:118]: parameter formatting.

File: presto-spark-classloader-interface/src/main/java/com/facebook/presto/spark/classloader_interface/PrestoSparkShuffleDescriptor.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * PrestoSparkShuffleDescriptor is used as a container to carry over Spark shuffle related meta information
+ * (e.g. ShuffleHandle, number of partitions etc.) from Spark related structures (Spark Rdd, Spark Partition etc.)
+ * to Presto-Spark land. These meta information are essential to delegate the shuffle operation from JVM to external
+ * native execution process.
+ */
+public abstract class PrestoSparkShuffleDescriptor
```
Comment (arunthirupathi) [inline presto-spark-classloader-interface/src/main/java/com/facebook/presto/spark/classloader_interface/PrestoSparkShuffleDescriptor.java:27]: Why is this class marked as abstract and has no abstract methods ?

File: presto-spark-classloader-interface/src/main/java/com/facebook/presto/spark/classloader_interface/PrestoSparkShuffleWriteDescriptor.java [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spark.classloader_interface;
+
+import org.apache.spark.shuffle.ShuffleHandle;
+
+public class PrestoSparkShuffleWriteDescriptor
```
Comment (arunthirupathi) [inline presto-spark-classloader-interface/src/main/java/com/facebook/presto/spark/classloader_interface/PrestoSparkShuffleWriteDescriptor.java:18]: The class hierarchy looks hacky.

**Review by arunthirupathi — APPROVED** [review]
Left some minor comments, once addressed good to go.

File: presto-spark-classloader-interface/src/main/java/com/facebook/presto/spark/classloader_interface/PrestoSparkShuffleWriteDescriptor.java (hunk 5 above)
Comment (tanjialiang) [reply presto-spark-classloader-interface/src/main/java/com/facebook/presto/spark/classloader_interface/PrestoSparkShuffleWriteDescriptor.java:18]: Yeah the hierarchy is hacky I think. Let me see what I can do.

**Pull request merged**